### ISSUE TITLE
augmentation process and focal loss

### ISSUE THREAD

**Issue by DP1701** [ultralytics/yolov5#6503]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello all,
is it possible to see how many images the dataset has grown by after the augmentation process?

**Comment by glenn-jocher** [comment]
@DP1701 👋 Hello! Thanks for asking about **image augmentation**. YOLOv5 🚀 applies online imagespace and colorspace augmentations in the trainloader (but not the val_loader) to present a new and unique augmented [Mosaic](https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/utils/datasets.py#L678-L732) (original image + 3 random images) each time an image is loaded for training. **Images are never presented twice in the same way**.

![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/120995721-f3cfed00-c785-11eb-8ee2-b6ef2fa205e8.jpg)

### Augmentation Hyperparameters

The hyperparameters used to define these augmentations are in your hyperparameter file (default `data/hyp.scratch.yaml`) defined when training:
```
python train.py --hyp hyp.scratch.yaml
```
https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/data/hyp.scratch.yaml#L1-L33


### Augmentation Previews

You can view the effect of your augmentation policy in your train_batch*.jpg images once training starts. These images will be in your train logging directory, typically `yolov5/runs/train/exp`:

`train_batch0.jpg` shows train batch 0 mosaics and labels:  
<img src="https://user-images.githubusercontent.com/26833433/131255960-b536647f-7c61-4f60-bbc5-cb2544d71b2a.jpg" width="800">


### YOLOv5 Albumentations Integration

[YOLOv5](https://github.com/ultralytics/yolov5) 🚀 is now fully integrated with [Albumentations](https://github.com/albumentations-team/albumentations), a popular open-source image augmentation package. Now you can train the world's best Vision AI models even better with custom Albumentations 😃! 

PR https://github.com/ultralytics/yolov5/pull/3882 implements this integration, which will automatically apply Albumentations transforms during YOLOv5 training if `albumentations>=1.0.3` is installed in your environment. See https://github.com/ultralytics/yolov5/pull/3882 for full details.

Example `train_batch0.jpg` on COCO128 dataset with Blur, MedianBlur and ToGray. See the YOLOv5 [Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) to reproduce: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img src="https://user-images.githubusercontent.com/26833433/124400879-ff331b80-dd25-11eb-9b67-fe85ac4ca104.jpg" width="800">

Good luck 🍀 and let us know if you have any other questions!

**Comment by DP1701** [comment]
@glenn-jocher Thank you for the detailed explanation. This helps!

However, I still have a question: What do the values (fraction) for HSV-Hue, HSV-Saturation and HSV-Value augmentation mean? 
Unfortunately, I have not yet found an answer to this question.

**Comment by glenn-jocher** [comment]
@DP1701 these are fractional gains on each of the image HSV channels. The augmentation function in here:
https://github.com/ultralytics/yolov5/blob/4c409332667477560200958b513b958bb8fdef71/utils/augmentations.py#L47-L61

**Comment by DP1701** [comment]
@glenn-jocher 
Thank you for the explanation and the hint.

I have another question about focal loss. In the hyperparameterset I can change the gamma value, but where can I set the alpha values?

**Comment by glenn-jocher** [comment]
@DP1701 Focal Loss function with `alpha` is here:
https://github.com/ultralytics/yolov5/blob/9bc72a3ac2253b96b006baaacf099ca4ace31a61/utils/loss.py#L35-L63

**Comment by DP1701** [comment]
@glenn-jocher 
Ok, if I set the gamma value in the [hyperparameterset](https://github.com/ultralytics/yolov5/blob/9bc72a3ac2253b96b006baaacf099ca4ace31a61/data/hyps/hyp.scratch.yaml#L21) to 0, then the default value of 1.5 is set to 0. But I have to change the alpha value directly in the function in line [37](https://github.com/ultralytics/yolov5/blob/9bc72a3ac2253b96b006baaacf099ca4ace31a61/utils/loss.py#L37)?

**Comment by glenn-jocher** [comment]
@DP1701 yes you just update alpha in the code if you want, it's not a hyperparameter.

**Comment by DP1701** [comment]
@glenn-jocher Thanks for the clarification.

I have an unbalanced data set. So the option I think will help to compensate this. Or is there another better option?

**Comment by Akshaysharma29** [comment]
Hi @glenn-jocher thanks for the explanations. I have two question 
1. if I do degrees: 90  in this file [yolov5/data/hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/data/hyp.scratch.yaml#L1-L33)  then the image will rotate from -90 to +90?
2. What will be the probability of rotation?

**Comment by glenn-jocher** [comment]
@Akshaysharma29 yes, probability is 1.0

**Comment by Akshaysharma29** [comment]
@glenn-jocher thanks for the response🙂. what should I do if I only required 0,90,180 rotation, not any other rotation?
any suggestion?

**Comment by glenn-jocher** [comment]
@Akshaysharma29 uncomment L145 here:
https://github.com/ultralytics/yolov5/blob/07221f1591c0ce7544bd5abf343dac2987e1614e/utils/augmentations.py#L142-L149
